kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/MetadataCleanup.java

```diff
@@ -0,0 +1,198 @@
+                    + "file that indicates that the log files are no longer needed. ",
+                tablePath,
+                potentialLogFilesToDelete.size());
+          }
+          break;
+        }
+
+        if (FileNames.isCommitFile(nextFile.getPath())) {
+          // Step 3: If the file is a delta log file, add it to the `potentialFilesToDelete` list
+          // We can't delete these files until we encounter a checkpoint later that indicates
+          // that the log files are no longer needed.
+          potentialLogFilesToDelete.add(nextFile.getPath());
+        } else if (FileNames.isCheckpointFile(nextFile.getPath())) {
+          // Step 4: If the file is a checkpoint file, add it to the `lastSeenCheckpointFiles`
+          long newLastSeenCheckpointVersion = FileNames.checkpointVersion(nextFile.getPath());
+          checkArgument(
+              lastSeenCheckpointVersion == -1
+                  || newLastSeenCheckpointVersion >= lastSeenCheckpointVersion);
+
+          if (lastSeenCheckpointVersion != -1
+              && newLastSeenCheckpointVersion > lastSeenCheckpointVersion) {
+            // We have found checkpoint file for a new version. This means the files gathered for
+            // the last checkpoint version are not complete (most likely an incomplete multipart
+            // checkpoint). We should delete the files gathered so far and start fresh
+            // last seen checkpoint state
+            logger.info(
+                "{}: Incomplete checkpoint files found at version {}, ignoring the checkpoint"
+                    + " files and adding them to potential log file delete list",
+                tablePath,
+                lastSeenCheckpointVersion);
+            potentialLogFilesToDelete.addAll(lastSeenCheckpointFiles);
+            lastSeenCheckpointFiles.clear();
+          }
+
+          lastSeenCheckpointFiles.add(nextFile.getPath());
+          lastSeenCheckpointVersion = newLastSeenCheckpointVersion;
+        }
+        // TODO: do we need to delete unknown file types?
+      }
+    }
+    logger.info("{}: Deleted {} log files older than {}", tablePath, numDeleted, fileCutOffTime);
+  }
+
+  private static CloseableIterator<FileStatus> listDeltaLogs(Engine engine, Path tablePath)
+      throws IOException {
+    Path logPath = new Path(tablePath, "_delta_log");
+    // TODO: Currently we don't update the timestamps of files to be monotonically increasing.
+    // In future we can do something similar to Delta Spark to make the timestamps monotonically
+    // increasing. See `BufferingLogDeletionIterator` in Delta Spark.
+    return engine.getFileSystemClient().listFrom(FileNames.listingPrefix(logPath, 0));
+  }
+
+  private static int deleteLogFiles(Engine engine, List<String> logFiles) throws IOException {
+    int numDeleted = 0;
+    for (String logFile : logFiles) {
+      if (!wrapEngineExceptionThrowsIO(
```

Comment:
why the `!`? `FileSystemClient::delete says it @return true if delete is successful else false.`

so if the delete is successful, it returns true, and we want to increment numDeleted

Reply:
and we should ensure we have tests for this too, btw